src/riak_kv_qry_worker.erl

```diff
@@ -339,6 +351,20 @@ run_select_on_chunk(SubQId, Chunk, #state{qry = Query,
             run_select_on_group(Query, SelClause, DecodedChunk, QueryResult1)
     end.
 
+%% ------------------------------------------------------------
+%% Helper function to return decoded query results for the current
+%% Chunk:
+%%
+%%   if already decoded, simply returns the decoded data
+%%
+%%   if not, decodes and returns
+%% ------------------------------------------------------------
+
+get_decoded_results({decoded, Chunk}) ->
+    Chunk;
+get_decoded_results(Chunk) ->
+    decode_results(lists:flatten(Chunk)).
```

Comment:
I tried removing both instances of `lists:flatten/1` in this PR, and of course it doesn't work! We get this error:
```
2016-11-17 09:55:15.258 [error] <0.1541.0> CRASH REPORT Process <0.1541.0> with 0 neighbours exited with reason: no function clause matching riak_kv_qry_worker:decode_results([[{<<16,0,0,0,3,12,183,128,8,16,0,0,0,2,18,179,220,173,247,91,133,210,195,98,153,0,8,18,179,...>>,...},...]]) line 202 in gen_server:terminate/6 line 744
```
It seems like in `riak_kv_fold_buffer` a chunk of rows called an `Item` is added to an accumulator but in a non-flat way, e.g. `[Row1,Row2]|Acc]`. This is interesting because when `riak_kv_fold_buffer` works out the size of the buffer it uses the number of `Item` chunks it has received, and not the number of deep rows.

If you thought the flatten was a problem we could rewrite `decode_results/1` to iterate a deep list so flattening is not necessary. Flattening basically reallocates the list, each element requires one word of memory plus one word for each element: http://erlang.org/doc/efficiency_guide/advanced.html.